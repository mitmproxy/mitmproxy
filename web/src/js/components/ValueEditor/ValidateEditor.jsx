import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ValueEditor from './ValueEditor'
import classnames from 'classnames'


export default class ValidateEditor extends Component {

    static propTypes = {
        content: PropTypes.string.isRequired,
        readonly: PropTypes.bool,
        onDone: PropTypes.func.isRequired,
        className: PropTypes.string,
        isValid: PropTypes.func.isRequired,
    }

    constructor(props) {
        super(props)
        this.state = { valid: props.isValid(props.content) }
        this.onInput = this.onInput.bind(this)
        this.onDone = this.onDone.bind(this)
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        this.setState({ valid: nextProps.isValid(nextProps.content) })
    }

    onInput(content) {
        this.setState({ valid: this.props.isValid(content) })
    }

    onDone(content) {
        if (!this.props.isValid(content)) {
            this.editor.reset()
            content = this.props.content
        }
        this.props.onDone(content)
    }

    render() {
        let className = classnames(
            this.props.className,
            {
                'has-success': this.state.valid,
                'has-warning': !this.state.valid
            }
        )
        return (
            <ValueEditor
                content={this.props.content}
                readonly={this.props.readonly}
                onDone={this.onDone}
                onInput={this.onInput}
                className={className}
                ref={(e) => this.editor = e}
            />
        )
    }
}
