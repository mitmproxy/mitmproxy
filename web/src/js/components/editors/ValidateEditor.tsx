import React, { useRef, useState } from "react";
import ValueEditor, { ValueEditorProps } from "./ValueEditor";
import classnames from "classnames";

interface ValidateEditorProps extends ValueEditorProps {
    isValid: (content: string) => boolean;
}

export default function ValidateEditor(props: ValidateEditorProps) {
    const [isValid, setIsValid] = useState(props.isValid(props.content));
    const editor = useRef<ValueEditor>(null);

    const onEditDone = (newVal: string) => {
        if (props.isValid(newVal)) {
            props.onEditDone(newVal);
        } else {
            editor.current?.resetValue();
        }
    };

    const className = classnames(
        props.className,
        isValid ? "has-success" : "has-warning"
    );
    return (
        <ValueEditor
            {...props}
            className={className}
            onInput={(newVal) => setIsValid(props.isValid(newVal))}
            onEditDone={onEditDone}
            ref={editor}
        />
    );
}
