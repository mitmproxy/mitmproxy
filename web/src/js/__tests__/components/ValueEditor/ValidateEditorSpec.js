import React from 'react'
import renderer from 'react-test-renderer'
import TestUtils from 'react-dom/test-utils'
import ValidateEditor from '../../../components/ValueEditor/ValidateEditor'

describe('ValidateEditor Component', () => {
    let validateFn = jest.fn( content => content.length == 3),
        doneFn = jest.fn()

    it('should render correctly', () => {
        let validateEditor = renderer.create(
            <ValidateEditor content="foo" onDone={doneFn} isValid={validateFn}/>
        ),
            tree = validateEditor.toJSON()
        expect(tree).toMatchSnapshot()
    })

    let validateEditor = TestUtils.renderIntoDocument(
        <ValidateEditor content="foo" onDone={doneFn} isValid={validateFn}/>
    )
    it('should handle componentWillReceiveProps', () => {
        let mockProps = {
            isValid: s => s.length == 3,
            content: "bar"
        }
        validateEditor.UNSAFE_componentWillReceiveProps(mockProps)
        expect(validateEditor.state.valid).toBeTruthy()
        validateEditor.UNSAFE_componentWillReceiveProps({...mockProps, content: "bars"})
        expect(validateEditor.state.valid).toBeFalsy()

    })

    it('should handle input', () => {
        validateEditor.onInput("foo bar")
        expect(validateFn).toBeCalledWith("foo bar")
    })

    it('should handle done', () => {
        // invalid
        validateEditor.editor.reset = jest.fn()
        validateEditor.onDone("foo bar")
        expect(validateEditor.editor.reset).toBeCalled()
        // valid
        validateEditor.onDone("bar")
        expect(doneFn).toBeCalledWith("bar")
    })
})
