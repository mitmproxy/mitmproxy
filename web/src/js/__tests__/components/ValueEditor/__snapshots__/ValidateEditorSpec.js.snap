// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidateEditor Component should render correctly 1`] = `
<div
  className="inline-input editable has-success"
  dangerouslySetInnerHTML={
    Object {
      "__html": "foo",
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onInput={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  tabIndex={0}
/>
`;
